File: ЯПlab1/ui.c
#include "ui.h"
#include <stdio.h>

void ui_print_menu(void) {
    printf("\n=== CONTACT LIST ===\n");
    printf("1. Print all contacts\n");
    printf("2. Search for a contact\n");
    printf("3. Add new contact\n");
    printf("4. Exit\n");
    printf("Choose action: ");
}


void ui_show_all_contacts(ContactBST* tree) {
    if (tree == NULL || tree->root==NULL) {
        printf("The phone book is empty\n");
        return;
    }

    printf("\n--- All Contacts ---\n");
    bst_inorder(tree, contact_print);
}


void ui_search_contact(ContactBST* tree) {
    if (tree == NULL) {
        printf("Error: the tree is not initialized\n");
        return;
    }

    int choice;
    printf("\n=== CONTACT SEARCH ===\n");
    printf("1. Search by name\n");
    printf("2. Search by surname\n");
    printf("3. Search by name and surname\n");
    printf("4. Search by phone\n");
    printf("5. Search by email\n");
    printf("Choose search type: ");
    scanf("%d", &choice);
    getchar();

    if (choice == 1) {
        char name[MAX_NAME];
        printf("Enter name: ");
        scanf("%49s", name);
        printf("\nSearch results:\n");

        bst_search_all_fields(tree,name, contact_print);
    } else if (choice == 2) {
        char surname[MAX_SURNAME];
        printf("Enter surname: ");
        scanf("%49s", surname);
        printf("\nSearch results:\n");
        bst_search_all_fields(tree,surname, contact_print);
    } else if (choice == 3) {
        char name[MAX_NAME], surname[MAX_NAME];
        printf("Enter name: ");
        scanf("%49s", name);
        printf("Enter surname: ");
        scanf("%49s", surname);

        printf("\nSearch results:\n");
        bst_search_by_name_and_surname_wrapper(tree, name, surname, contact_print);
    } else if (choice == 4) {
        char phone[MAX_PHONE];
        printf("Enter phone: ");
        scanf("%49s", phone);
        printf("\nSearch results:\n");
        bst_search_all_fields(tree, phone, contact_print);
    } else if (choice == 5) {
        char email[MAX_EMAIL];
        printf("Enter email: ");
        scanf("%49s", email);
        printf("\nSearch results:\n");
        bst_search_all_fields(tree, email, contact_print);
    }
}

void ui_add_contact(ContactBST* tree) {
    if (tree == NULL) {
        printf("Error: the tree is not initialized\n");
        return;
    }

    char name[MAX_NAME], surname[MAX_SURNAME], phone[MAX_PHONE], email[MAX_EMAIL];

    printf("Enter name: ");
    scanf("%49s", name);
    printf("Enter surname: ");
    scanf("%49s", surname);
    printf("Enter phone: ");
    scanf("%49s", phone);
    printf("Enter email: ");
    scanf("%49s", email);

    Contact new_contact = contact_create(name, surname, phone, email);

    if (bst_insert(tree, new_contact)) {
        printf("Contact added successfully!\n");
    } else {
        printf("Error: failed to add contact\n");
    }
}

void ui_run(ContactBST* tree) {
    if (tree == NULL) {
        printf("Error: tree not initialized\n");
        return;
    }

    int choice;

    do {
        ui_print_menu();
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                ui_show_all_contacts(tree);
                break;
            case 2:
                ui_search_contact(tree);
                break;
            case 3:
                ui_add_contact(tree);
                break;
            case 4:
                printf("Exit...\n");
                break;
            default:
                printf("Invalid choice!\n");
        }
    } while (choice != 4);
}

File: ЯПlab2/ConsoleApplication2/UI.cs
using System;

namespace ConsoleApplication2
{
    public class UI
    {
        public static void PrintMenu()
        {
            Console.WriteLine("\n=== CONTACT LIST ===");
            Console.WriteLine("1. Print all contacts");
            Console.WriteLine("2. Search for a contact");
            Console.WriteLine("3. Add new contact");
            Console.WriteLine("4. Exit");
            Console.Write("Choose action: ");
        }

        public static void ShowAllContacts(ContactBST tree)
        {
            if (tree == null)
            {
                Console.WriteLine("The phone book is empty");
                return;
            }

            Console.WriteLine("\n--- All Contacts ---");
            tree.Inorder(contact => contact.Print());
        }


        public static void SearchContact(ContactBST tree)
        {
            if (tree == null)
            {
                Console.WriteLine("Error: the tree is not initialized");
                return;
            }

            Console.WriteLine("\n=== CONTACT SEARCH ===");
            Console.WriteLine("1. Search by name");
            Console.WriteLine("2. Search by surname");
            Console.WriteLine("3. Search by name and surname");
            Console.WriteLine("4. Search by phone");
            Console.WriteLine("5. Search by email");
            Console.Write("Choose search type: ");

            if (!int.TryParse(Console.ReadLine(), out int choice))
            {
                Console.WriteLine("Invalid input!");
                return;
            }

            switch (choice)
            {
                case 1:
                    Console.Write("Enter name: ");
                    string name = Console.ReadLine();
                    Console.WriteLine("\nSearch results:");
                    tree.SearchAllFields(name, contact => contact.Print());
                    break;

                case 2:
                    Console.Write("Enter surname: ");
                    string surname = Console.ReadLine();
                    Console.WriteLine("\nSearch results:");
                    tree.SearchAllFields(surname, contact => contact.Print());
                    break;

                case 3:
                    Console.Write("Enter name: ");
                    string name3 = Console.ReadLine();
                    Console.Write("Enter surname: ");
                    string surname3 = Console.ReadLine();
                    Console.WriteLine("\nSearch results:");
                    tree.SearchByNameAndSurname(name3, surname3, contact => contact.Print());
                    break;

                case 4:
                    Console.Write("Enter phone: ");
                    string phone = Console.ReadLine();
                    Console.WriteLine("\nSearch results:");
                    tree.SearchAllFields(phone, contact => contact.Print());
                    break;

                case 5:
                    Console.Write("Enter email: ");
                    string email = Console.ReadLine();
                    Console.WriteLine("\nSearch results:");
                    tree.SearchAllFields(email, contact => contact.Print());
                    break;

                default:
                    Console.WriteLine("Invalid choice!");
                    break;
            }
        }

        public static void AddContact(ContactBST tree)
        {
            if (tree == null)
            {
                Console.WriteLine("Error: the tree is not initialized");
                return;
            }

            Console.Write("Enter name: ");
            string name = Console.ReadLine();
            Console.Write("Enter surname: ");
            string surname = Console.ReadLine();
            Console.Write("Enter phone: ");
            string phone = Console.ReadLine();
            Console.Write("Enter email: ");
            string email = Console.ReadLine();

            Contact newContact = Contact.Create(name, surname, phone, email);

            if (tree.Insert(newContact))
            {
                Console.WriteLine("Contact added successfully!");
                FileStorage.Save(tree);
            }
            else
            {
                Console.WriteLine("Error: failed to add contact (possibly duplicate)");
            }
        }


        public static void run(ContactBST tree)
        {
            if (tree == null)
            {
                Console.WriteLine("Error: the tree is not initialized");
                return;
            }

            int choice;
            do
            {
                PrintMenu();
                if (!int.TryParse(Console.ReadLine(), out choice))
                {
                    Console.WriteLine("Invalid input!");
                    continue;
                }

                switch (choice)
                {
                    case 1:
                        ShowAllContacts(tree);
                        break;
                    case 2:
                        SearchContact(tree);
                        break;
                    case 3:
                        AddContact(tree);
                        break;
                    case 4:
                        Console.WriteLine("Exit...");
                        break;
                    default:
                        Console.WriteLine("Invalid choice!");
                        break;
                }
            } while (choice != 4);
        }
    }
}